File: examples/perl/mega_flow/pipeline.pl
package Pipeline;
# Pipeline — exercises Perl's idiomatic flow constructs: refs,
# map / grep / sort, anonymous subs, for loops, eval-blocks,
# ternary, regex, dispatch tables.
use strict; use warnings;
require "./storage.pl";

sub StorePersist {
    my ($valid) = @_;
    return Storage::persist($valid);
}

sub orchestrate {
    my ($envelope) = @_;
    my $cmd  = $envelope->{cmd};
    my $user = $envelope->{user};

    # for loop + grep + map — filter and normalise tainted tokens.
    my @tokens;
    for my $part (split /\s+/, $cmd) {
        push @tokens, $part if length $part;
    }

    @tokens = map { my $s = $_; $s =~ s/^\s+|\s+$//g; $s } @tokens;
    @tokens = grep { length } @tokens;

    # Reduce with anonymous-sub closure — taint rides the accumulator.
    my $joiner = sub {
        my ($acc, $tok) = @_;
        return $acc eq '' ? $tok : "$acc $tok";
    };
    my $joined = '';
    for my $t (@tokens) {
        $joined = $joiner->($joined, $t);
    }

    # Dispatch via a hash of anonymous subs.
    my %routes = (
        run  => sub { my ($s) = @_; "$s" },
        eval => sub { my ($s) = @_; $s =~ s/^\s+|\s+$//g; $s },
    );
    my $kind = $envelope->{kind} // 'run';
    my $routed = exists $routes{$kind} ? $routes{$kind}->($joined) : $joined;

    # eval-block (Perl's try/catch) — taint survives the branch.
    my $valid;
    eval {
        die "empty\n" unless length $routed;
        $valid = { %$envelope, cmd => $routed, user => $user };
        1;
    } or do {
        $valid = { %$envelope, cmd => $routed // '', user => $user };
    };

    return StorePersist($valid);
}
1;

File: examples/lua/mega_flow/pipeline.lua
-- Pipeline — exercises Lua's idiomatic flow constructs: closures,
-- multiple returns, varargs, numeric / generic for, coroutines,
-- string.gmatch iterators, pcall (try/catch), metatables, dispatch
-- tables, and chained string.* calls.
local Storage = require("storage")
local M = {}

-- Coroutine producer — yields every tainted token to the caller.
local function tokenize(cmd)
    return coroutine.wrap(function()
        for word in string.gmatch(cmd, "%S+") do
            coroutine.yield(word)
        end
    end)
end

-- Closure factory — returns a reducer that joins tokens with `sep`.
local function make_joiner(sep)
    return function(acc, tok)
        if acc == "" then
            return tok
        else
            return acc .. sep .. tok
        end
    end
end

function M.orchestrate(envelope, ...)
    local extra = { ... }
    local cmd = envelope.cmd
    local user = envelope.user

    -- Generic-for iterator over the coroutine.
    local parts = {}
    for word in tokenize(cmd) do
        parts[#parts + 1] = word
    end

    -- Numeric-for + closure reduce — taint rides the accumulator.
    local joiner = make_joiner(" ")
    local joined = ""
    for i = 1, #parts do
        joined = joiner(joined, parts[i])
    end

    -- Dispatch table (Lua's pattern-match analogue).
    local routes = {
        run  = function(s) return "" .. s end,
        ["eval"] = function(s) return s:gsub("^%s+", ""):gsub("%s+$", "") end,
    }
    local route = routes[envelope.kind] or function(s) return s end
    local routed = route(joined)

    -- pcall (Lua's try/catch) — taint survives the branch.
    local ok, valid = pcall(function()
        assert(#routed > 0, "empty")
        return { kind = envelope.kind, cmd = routed, user = user, length = #routed, extras = extra or envelope.extras }
    end)
    if not ok then
        valid = { kind = envelope.kind, cmd = routed or "", user = user, length = 0, extras = {} }
    end
    setmetatable(valid, { __index = { cmd_value = function(self) return self.cmd end } })

    return Storage.persist(valid)
end

return M
